added support for the schematics agent policies

rajatagarwal-ibm opened the pull request:
### Description

- Added a new submodule for the schematics agent policy handling.
- Updated the kubernetes example to create policies

### Release required?
<!--- Identify the type of release. For information about the changes in a semantic versioning release, see [Release versioning](https://terraform-ibm-modules.github.io/documentation/#/versioning). --->

- [ ] No release
- [ ] Patch release (`x.x.X`)
- [x] Minor release (`x.X.x`)
- [ ] Major release (`X.x.x`)

##### Release notes content

<!--- If a release is required, replace this text with information that users need to know about the release. Write the release notes to help users understand the changes, and include information about how to update from the previous version.

Your notes help the merger write the commit message for the PR that is published in the release notes for the module. --->

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### For mergers

- Use a conventional commit message to set the release level. Follow the [guidelines](https://terraform-ibm-modules.github.io/documentation/#/merging.md).
- Include information that users need to know about the PR in the commit message. The commit message becomes part of the GitHub release notes.
- Use the **Squash and merge** option.

jor2 commented on examples/kubernetes/main.tf:
```diff
@@ -100,4 +100,30 @@ module "schematics_agent" {
+
+      target = [{
+        selector_kind = "ids"
+        selector_ids  = [module.schematics_agent.agent_id]
```
circular dependency?

jor2 commented on main.tf:
```diff
@@ -70,3 +70,8 @@ resource "null_resource" "agent_deployment_status" {
   }
 }
+
+module "schematics_policy" {
```
module doesn't depend on the agent being deployed. If a policy references an agent ID, the agent should exist first.

rajatagarwal-ibm replied:
yes, this module doesn't depend on the agent being deployed. What it expects is the agent_id if you want to pass, but that's not the mandatory requirement.

jor2 commented on variables.tf:
```diff
@@ -123,3 +123,9 @@ variable "disable_agent" {
+
+variable "schematics_policies" {
+  description = "Schematics agent policies to create"
+  type        = any
```
should this be more specific?

rajatagarwal-ibm replied:
Either I have to duplicate the schema defined in the submodule, or to use `any`. I use `any` as all the schema and validation handled in the submodule.

rajatagarwal-ibm commented on modules/schematics-policy/variables.tf:
```diff
@@ -0,0 +1,147 @@
+variable "schematics_policies" {
+  description = "Schematics agent policies to create"
+  type = map(object({
+    name           = string
```
I don't think so, because if you pass an empty string, then the provider would complain too.

rajatagarwal-ibm commented:
/run pipeline

rajatagarwal-ibm commented:
/run pipeline

ocofaigh commented on variables.tf:
```diff
@@ -123,3 +123,9 @@ variable "disable_agent" {
+
+variable "schematics_policies" {
+  description = "Schematics agent policies to create"
+  type        = any
```
we should not use `any` when we know the schema. Please add the full schema here. We only use `any` if we do not know the schema. 

I'm also wondering should this not be a list(object) type? A list of policies? The default value would indicate its a map. You probably also need to set this to `nullable = false`

ocofaigh commented on modules/schematics-policy/README.md:
```diff
@@ -0,0 +1,3 @@
+# Schematics Agent Policy
+
+This submodule creates and configures the Schematics Agent Policy. More info on working with the agent policies can be found [here](https://cloud.ibm.com/docs/schematics?topic=schematics-policy-manage&interface=ui).
```
This readme is for a module and therefore should have all the associated info on module (usage, required permissions, the terradocs info etc)

rajatagarwal-ibm commented on modules/schematics-policy/variables.tf:
```diff
@@ -0,0 +1,147 @@
+  validation {
+    condition = alltrue([
+      for _, p in var.schematics_policies : (
+      try(p.location, null) == null || contains(["us-south", "us-east", "eu-gb", "eu-de"], try(p.location, null)))
```
It doesn't support `eu-fr2`.

ocofaigh reviewed (commented):
@rajatagarwal-ibm A test we need to do is what happens if you have 3 policies, and you remove the 2nd one. Does it destroy and recreate policies due to the index changing? If so we need to add logic to prevent that. if you recall we have hit that in several modules to date and its always a breaking change to fix it

ocofaigh commented on variables.tf:
```diff
@@ -123,3 +123,151 @@ variable "disable_agent" {
+          for ap in try(param.agent_assignment_policy_parameter, []) : [
+            for scope in try(ap.selector_scope, []) : alltrue([
+              for loc in try(scope.locations, []) :
+              contains(["us-south", "us-east", "eu-gb", "eu-de", "ca-mon", "ca-tor"], loc)
```
please add back eu-fr2

ocofaigh reviewed (changes requested):
See latest comments. Can you add back eu-fr2 please? It is an allowlisted only region, but module needs to support it for accounts that are allowlisted

rajatagarwal-ibm commented:
OpenShift test failed again with the same error. 

I just saw there's another line in the logs 

```
2025/12/22 17:00:52 Terraform apply | module.ocp_base.null_resource.confirm_network_healthy[0] (local-exec): Running script to ensure kube master can communicate with all worker nodes..
2025/12/22 17:00:52 Terraform apply | module.ocp_base.null_resource.confirm_network_healthy[0] (local-exec): Error from server (Forbidden): pods is forbidden: User "IAM#goldeneye.development@ibm.com" cannot list resource "pods" in API group "" in the namespace "calico-system"
2025/12/22 17:00:52 Terraform apply | module.ocp_base.null_resource.confirm_network_healthy[0] (local-exec): No calico-node pods found. Retrying in 10s. (Attempt 1 / 10)
```

I have no clue about this right now, I will investigate it later.

rajatagarwal-ibm commented:
/run pipeline

ocofaigh commented:
@rajatagarwal-ibm The above logs you printed are not the reason for the test failure. The confirm_network_healthy script is supposed to log this on first attempt, but as you can see retries 10 times, and it usually passes within those retries.
The reason for the test failure in [this run ](https://github.com/terraform-ibm-modules/terraform-ibm-schematics-agent/actions/runs/20466465298/job/58811663160) was:
```
         2025/12/23 16:54:54 Terraform plan | Error: Invalid value for variable
         2025/12/23 16:54:54 Terraform plan | 
         2025/12/23 16:54:54 Terraform plan |   on main.tf line 116, in module "schematics_agent":
         2025/12/23 16:54:54 Terraform plan |     ├────────────────
         2025/12/23 16:54:54 Terraform plan |     │ var.schematics_policies is map of object with 1 element
```

The test also failed with:
```
         2025/12/23 17:23:33 Terraform apply | Error: RequestError: send request failed
         2025/12/23 17:23:33 Terraform apply | caused by: Put "https://s3.eu-fr2.cloud-object-storage.appdomain.cloud/sa-k8s-kth-bucket-l9yc": dial tcp: lookup s3.eu-fr2.cloud-object-storage.appdomain.cloud on 172.21.0.10:53: no such host
```
Please do not have automated tests running in `eu-fr2` region

ocofaigh commented on tests/pr_test.go:
```diff
@@ -21,6 +21,7 @@ var validRegions = []string{
 	"us-east",
 	"ca-mon",
 	"ca-tor",
+	"eu-fr2",
```
remove this - we should not be doing automated tests in this region. Manual test is ok

rajatagarwal-ibm commented:
I have updated the PR with the following changes:
- [x] Removed `"eu-fr2"` region from the automated tests
- [x] Added "eu-fr2" region in the module
- [x] For additional clarity, renamed "TestRunUpgradeSchematics" to "TestRunOpenShiftUpgradeSchematics".
- [x] Updated the logic such that it would always create a new RG for "TestRunOpenShiftExampleInSchematics" and "TestRunOpenShiftUpgradeSchematics". Added a comment too. (_**Note**: "TestRunKubernetesExampleInSchematics" test would use the existing logic to reuse the same RG, as that example does not create the Schematics Agent Policies._) 
- [x] Debug failures and updated validation accordingly.

rajatagarwal-ibm commented:
/run pipeline

terraform-ibm-modules-ops commented:
:tada: This PR is included in version 1.2.0 :tada:

The release is available on [GitHub release](https://github.com/terraform-ibm-modules/terraform-ibm-schematics-agent/releases/tag/v1.2.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
